### PR TITLE
[Model] Add LingBot-Video support (MoE 30B + Dense 1.3B)

### PR DESCRIPTION
**Pull request by nholmber**
# Add LingBot-Video Support (MoE 30B + Dense 1.3B)

## Summary

Adds xDiT support for [LingBot-Video](https://github.com/Robbyant/lingbot-video), a Mixture-of-Experts video generation model (30B total / 3B active per token) for embodied intelligence. Supports both the MoE 30B and Dense 1.3B variants with T2V, TI2V, and refiner pipelines.

**Requires:** `pip install lingbot-video` (model not in diffusers, ships as standalone package)

## Features

- **Parallelism:** USP (Ulysses) u=2/4/8, CFG parallel, FSDP, parallel VAE
- **Quantization:** FP8 attention (AITER), FP8/FP4 GEMMs (torchao), hybrid attention schedule
- **torch.compile:** Zero graph breaks on both Dense and MoE models (2.13x over eager)
- **Modes:** T2V (`--prompt`), TI2V (`--input_images`), refiner (`--task refine`)
- **JSON prompts:** `--prompt path/to/prompt.json`; model requires structured JSON captions

## Usage

LingBot-Video requires structured JSON captions, not plain text prompts. The JSON
contains scene descriptions, camera info, and per-element actions with timestamps:

```json
{
  "caption": {
    "comprehensive_description": {
      "scene_content_description": "A young child is outdoors on a sunny day, playing with bubbles...",
      "camera_movement_description": "The camera is stationary, medium close-up, eye level..."
    },
    "camera_info": {
      "color": "Saturated", "frame_size": "Medium Close Up",
      "shot_type_angle": "Low angle", "lens_size": "Long Lens",
      "composition": "Center", "lighting": "Hard light", "lighting_type": "Daylight"
    },
    "prominent_elements": [
      {
        "name": "young child",
        "description": "A child with short brown hair wearing a colorful striped shirt",
        "actions": [
          {"timestamp": "[0.0s - 2.0s]", "action": "holds bubble wand and blows gently"},
          {"timestamp": "[2.0s - 5.0s]", "action": "waves the wand through the air"}
        ]
      }
    ]
  },
  "duration": 5
}
```

Use the [prompt rewriter](https://github.com/Robbyant/lingbot-video/tree/master/rewriter) to convert natural language to this format, or use the example prompts from `assets/cases/`.

```bash
# Dense 1.3B, single GPU
python -m xfuser.cli --model robbyant/lingbot-video-dense-1.3b \
  --prompt prompt.json --height 480 --width 832 --num_frames 81 \
  --guidance_scale 1.0 --use_torch_compile

# MoE 30B, 8 GPU with USP + CFG + FP8 attention + parallel VAE
python -m xfuser.cli --model robbyant/lingbot-video-moe-30b-a3b \
  --prompt prompt.json --height 1088 --width 1920 --num_frames 81 \
  --guidance_scale 3.0 --use_torch_compile --attention_backend AITER_FP8 \
  --ulysses_degree 4 --use_cfg_parallel --use_parallel_vae --nproc_per_node 8

# MoE 30B with refiner (480p base → 1080p output)
python -m xfuser.cli --model robbyant/lingbot-video-moe-30b-a3b \
  --prompt prompt.json --task refine \
  --height 480 --width 832 --num_frames 81 \
  --guidance_scale 1.0 --use_torch_compile --ulysses_degree 8 --nproc_per_node 8

# TI2V (text + image → video)
python -m xfuser.cli --model robbyant/lingbot-video-moe-30b-a3b \
  --prompt prompt.json --input_images first_frame.png \
  --height 480 --width 832 --num_frames 81 --guidance_scale 1.0 --use_torch_compile

# Refiner at 2K (--height/--width is the output resolution)
python -m xfuser.cli ... --task refine --height 1440 --width 2560
```

https://github.com/user-attachments/assets/24b8b688-8b1d-4fd2-9403-e498b035247d

https://github.com/user-attachments/assets/eeabdff4-16b2-40a6-a5b5-216fedf13419

### 480p Base (832×480, guidance_scale=1.0, 40 steps)

| Config | GPUs | Time | Speedup |
|--------|:----:|:----:|:-------:|
| BF16 | 1 | 44.0s | 1.00x |
| FP8 attn | 1 | 34.9s | 1.26x |
| FP8 attn + FP4 GEMMs | 1 | 34.2s | 1.29x |
| BF16 + u=4 | 4 | 29.8s | 1.48x |
| FP8 attn + FP4 + u=4 + pvae | 4 | 18.5s | 2.38x |
| BF16 + u=8 + pvae | 8 | **18.1s** | **2.43x** |

### 1080p Native (1920×1088, guidance_scale=3.0, 40 steps, 8 GPU)

| Config | Time | Speedup |
|--------|:----:|:-------:|
| BF16 + u=8 | 181.4s | 1.00x |
| BF16 + u=4+cfg=2 | 171.6s | 1.06x |
| FP8 attn + u=8 | 118.3s | 1.53x |
| FP8 attn + u=4+cfg=2 | **109.7s** | **1.65x** |

### T2V 1080p Refiner (480p→1080p, 40+8 steps, compiled)

| Config | GPUs | Time | Speedup |
|--------|:----:|:----:|:-------:|
| BF16 | 1 | 358.5s | 1.00x |
| BF16 + u=8 | 8 | 88.0s | 4.07x |
| BF16 + u=8 + pvae | 8 | 67.4s | 5.32x |
| FP8 attn + u=8 + pvae | 8 | 53.1s | 6.75x |
| FP8 attn + u=4+cfg=2 + pvae | 8 | **50.0s** | **7.17x** |

### TI2V 1080p Refiner (480p→1080p, 40+8 steps, compiled)

| Config | GPUs | Time | Speedup |
|--------|:----:|:----:|:-------:|
| BF16 | 1 | 359.7s | 1.00x |
| BF16 + u=8 | 8 | 87.0s | 4.13x |
| FP8 attn + u=8 + pvae | 8 | 53.0s | 6.79x |
| FP8 attn + u=4+cfg=2 + pvae | 8 | **49.9s** | **7.21x** |

## Architecture Notes

LingBot-Video differs from Wan 2.1:
- **Standalone model** — `LingBotVideoTransformer3DModel` is not a subclass of `WanTransformer3DModel`
- **Self-attention only** — text embeddings concatenated with video tokens before layer loop, no cross-attention
- **MoE FFN** — 128 experts, top-8 routing, `torch._grouped_mm` backend
- **Custom scheduler** — vendored `FlowUniPCMultistepScheduler` (not in diffusers)
- **Refiner** — separate 48-layer MoE transformer for upscaling (480p base → 1080p/2K target)

### Implementation approach
- Transformer wrapper with USP attention installed at load time. 
- Custom forward that chunks sequence before temb6 computation to avoid full-sequence memory allocations at high resolution
- Pipeline wrapper with MRO-based parent class dispatch for T2V/TI2V compatibility
- Manual component loading (bypasses diffusers `model_index.json` custom class resolution)
- FSDP via LingBot's native `apply_fsdp_inference` with `ignored_params` for mixed FP32/BF16
- Parallel VAE via distvae `WanDecoderAdapter`
- Pre-transposed expert weights to eliminate per-call `.bfloat16().transpose()` copies

### torch.compile fixes (zero graph breaks)
1. **RoPE lazy rebuild:** patched `forward()` to skip data-dependent `max_vals >= axes_lens` check
2. **`bool(tensor)` padding check:** compile-friendly forward avoids `has_padding` branch
3. **MoE `_reorder_tokens`:** replaced `torch.where(flat_scores != 0)` (dynamic shape) with static `argsort`
4. **Pre-transposed expert weights:** replace `w1/w3/w2` in-place with transposed copies, bound method patch on `_run_grouped_experts`

## Known Limitations

- FP8/FP4 GEMMs (torchao) only quantize attention linears, not MoE expert weights. `torch._grouped_mm` does not support FP8 dtype

### CODE REVIEW

File: xfuser/model_executor/models/runner_models/lingbot_video.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+            log("distvae WanEncoderAdapter not available, skipping parallel VAE encoder.")
+        except Exception as e:
+            log(f"Failed to patch VAE encoder: {e}")
+
+
+@register_model("robbyant/lingbot-video-moe-30b-a3b")
+@register_model("LingBot-Video-MoE")
+class xFuserLingBotVideoMoEModel(xFuserModel):
+
+    def _calculate_hybrid_attention_step_multiplier(self, input_args: dict) -> int:
```
Comment (avjves) [inline xfuser/model_executor/models/runner_models/lingbot_video.py:127]: Is this correct? Wan needs it as it runs two transformer calls per step, but the CFG implementation in this PR at least only runs once.

File: xfuser/model_executor/models/runner_models/lingbot_video.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+        negative_prompt=DEFAULT_NEGATIVE_PROMPT,
+        guidance_scale=3.0,
+        flow_shift=3.0,
+    )
+    settings = ModelSettings(
+        model_name="robbyant/lingbot-video-moe-30b-a3b",
+        output_name="lingbot_video_moe",
+        model_output_type="video",
+        fps=24,
+        fp8_gemm_module_list=["transformer.blocks"],
```
Comment (avjves) [inline xfuser/model_executor/models/runner_models/lingbot_video.py:159]: These FP8/FP4 GEMM quants only apply to `self.pipe`. Not sure if the idea is to also apply them to the refiner or not - at least there are some quant dtype assignments in the refinder loading function.

File: xfuser/model_executor/models/runner_models/lingbot_video.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+            model_name, torch_dtype=torch.bfloat16, subfolder="text_encoder",
+        )
+        processor = Qwen3VLProcessor.from_pretrained(
+            model_name, subfolder="processor",
+        )
+        scheduler = FlowUniPCMultistepScheduler.from_pretrained(
+            model_name, subfolder="scheduler",
+        )
+        if use_i2v:
+            from lingbot_video.pipeline_lingbot_video_i2v import LingBotVideoImageToVideoPipeline
```
Comment (avjves) [inline xfuser/model_executor/models/runner_models/lingbot_video.py:306]: Only the T2V task gets the parallel CFG support - is this intended?

**Review by avjves — COMMENTED** [review]
Thanks!
Left a couple of small comments, though not sure if action is required. Can you check them?

File: xfuser/model_executor/models/runner_models/lingbot_video.py (hunk 2 above)
Comment (nholmber) [reply xfuser/model_executor/models/runner_models/lingbot_video.py:159]: Added quantization also to the refiner.  It quantizes `.blocks` specifically (not the full transformer) because non-block modules like `patch_embedder` crash with MXFP4. Impact is minor overall since we dont touch the MoE

File: xfuser/model_executor/models/runner_models/lingbot_video.py (hunk 3 above)
Comment (nholmber) [reply xfuser/model_executor/models/runner_models/lingbot_video.py:306]: No, fixed.

File: xfuser/model_executor/models/runner_models/lingbot_video.py (hunk 1 above)
Comment (nholmber) [reply xfuser/model_executor/models/runner_models/lingbot_video.py:127]: LingBot has one transformer, but the pipeline calls it twice per step when `guidance_scale > 1.0` (separate conditional + unconditional forwards, not batched). So multiplier=2 is correct for that case. However, with `use_cfg_parallel` each rank only runs one forward per step, so the multiplier should be 1. Fixed in 5821baa to return 2 only when `guidance_scale > 1.0 AND NOT use_cfg_parallel`.

**Review by avjves — APPROVED** [review]
